rfcs/RFC-0002-configuration.md

```diff
@@ -0,0 +1,350 @@
+
+These two things will make it hard long term for tasks like Synthetic Data Generation (SDG), Training, Evals, etc to be run in a repeatable and knowledgeable manner by end users. Llama Stack of today is built to be a "black box" to not expose sensitive configuration to the user. Exposing the current provider configuration to a user will help them understand what they will be running for various providers as functionality gets more complex (SDG, Evals, Training, etc). Additionally, allowing a user to apply parts of a config on top of a running stack as opposed to taking the stack down and having the admin apply a full run config again seems like a more sustainable workflow.
+
+A concrete example is modification of the served models, the vLLM endpoint, etc: which currently require stopping LLS and vLLM and spinning both back up with new config. this API would enable the LLS to stay up and the provider config be modified to represent the current state of vLLM.
+
+Not having the ability to swap in configuration based on the circumstances at run time will make the potential for hardware detection, hyperparameter application, and the general UX of LLS more difficult than today.
+
+This seems like a table stakes set of operations, and one we will likely find ourselves needing as the feature set expands.
+
+## Solution Overview
+
+1. New API route: /v1/configurations
+2. Register endpoint for configurations: /v1/configurations/register
+3. New inspect API: /v1/inspect/configurations
+4. New resources in the llama-stack-client-python (and the other clients?) to handle Configurations
+5. Llama-stack-client commands for configurations
+6. Upstream consensus on this new API
+
+
+## Demo Code Available for Testing
+
+I have been working on this on my forks of LLS and the python client. This API works and enables users to apply configuration and inspect it all with a running server.
+
+To test this out yourself you can look at my custom fork of LLS and the client both from the `config` branch:
+
+How-to: https://github.com/cdoern/llama-stack/blob/config/docs/config_api.md
+llama-stack: https://github.com/cdoern/llama-stack/tree/config
+llama-stack-client-python: https://github.com/cdoern/llama-stack-client-python/tree/config
+
+
+## How can someone use this?
+
+Here is a modified version of the SDK example script which changes the configuration of a server:
+
+
+```python
+import os
+import sys
+import json
+import yaml
+
+
+def create_library_client(template="ollama"):
+    from llama_stack import LlamaStackAsLibraryClient
+
+    client = LlamaStackAsLibraryClient(template)
+    client.initialize()
+    return client
+
+
+client = create_library_client()
+
+prov = client.providers.list()
+config = client.configurations.inspect()
+
+print("Old Config \n", yaml.dump(config, indent=2))
```

Comment:
There can also be sensitive configurations such as API keys so we probably want to categorize fields that require redactions.

Reply:
hopefully this will be handled by the `UserConfig` class. I would imagine a user should not be able to see any API keys in the returned config